From 61b1d102129990f6e903c6ddcf46c7d79d1a1e99 Mon Sep 17 00:00:00 2001
From: Rich Felker <dalias@aerifal.cx>
Date: Thu, 5 Feb 2015 23:34:27 -0500
Subject: [PATCH] fix failure of fchmodat to report EOPNOTSUPP in the race path

in the case where a non-symlink file was replaced by a symlink during
the fchmodat operation with AT_SYMLINK_NOFOLLOW, mode change on the
new symlink target was successfully suppressed, but the error was not
reported. instead, fchmodat simply returned 0.
---
 src/stat/fchmodat.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/stat/fchmodat.c b/src/stat/fchmodat.c
index a894cb6..d94667a 100644
--- a/src/stat/fchmodat.c
+++ b/src/stat/fchmodat.c
@@ -29,8 +29,10 @@ int fchmodat(int fd, const char *path, mode_t mode, int flag)
 
 	__procfdname(proc, fd2);
 	ret = __syscall(SYS_fstatat, AT_FDCWD, proc, &st, 0);
-	if (!ret && !S_ISLNK(st.st_mode))
-		ret = __syscall(SYS_fchmodat, AT_FDCWD, proc, mode);
+	if (!ret) {
+		if (S_ISLNK(st.st_mode)) ret = -EOPNOTSUPP;
+		else ret = __syscall(SYS_fchmodat, AT_FDCWD, proc, mode);
+	}
 
 	__syscall(SYS_close, fd2);
 	return __syscall_ret(ret);
-- 
2.3.3

